gapis/memory/pointer.go

```diff
@@ -52,6 +52,18 @@ type Pointer interface {
 	// ISlice returns a new Slice of elements based from this pointer using
 	// start and end indices.
 	ISlice(start, end uint64, m *device.MemoryLayout) Slice
+
+	// All pointers by definition also implement ReflectPointer
+	APointer() bool
+}
+
+// This is a helper interface, if you want your pointer to be reflected
+// then it must ALSO implement this interface. Since reflection is slow
+// having a much smaller interface to check is significantly better
+// We name this APointer since reflect.Implements checks functions
+// in alphabetical order, meaning this should get hit first (or close to).
+type ReflectPointer interface {
```

Comment:
So, if I understand this correctly: You've added `APointer` to `Pointer` so that `reflect.Implements` can early out quicker for non-pointer types?

Given that I see no direct use of `ReflectPointer` (aside from the interface check), why do you need it? Implementers of `Pointer` are required to implement `ReflectPointer` anyway.

Reply:
Unfortunately not quite.

https://github.com/golang/go/blob/master/src/reflect/type.go#L1531

Implements: loops over every method in V (the type) type to see if it matches T (the interface). Even though both lists are in alphabetical order, and the code relies on this fact, there is no short-circuit for when V does not contain a method that is alphabetically <= the next method on T, and so all methods on V will get checked even though there is no possibility of a match

The only short-circuit through that code is when V contains ALL of the methods in T, so if we check against ReflectedPointer, if we DO match, ReflectedPointer, (which is likely the first function match on matching types), then we stop checking.

Reply:
Gotcha.